ReplayEditor/replayengine.cpp

```diff
@@ -406,6 +406,248 @@ void Replay::place_marks_at(I64 mark_in, I64 mark_out)
+
+bool Replay::move_frame(SongTime_t time)
+{
+     
+    if (m_mark_in >= m_frames.size() or m_mark_in < 0)
+    {
+        return false;
+    } 
+
+    if (m_mark_in == 0)
+    {
+        time = std::clamp(time, 1, -1 - m_frames[m_mark_in].ms + m_frames[m_mark_in + 1].ms);
+        
+    } 
+    else if (m_mark_in == m_frames.size() - 1)
+    {
+        time = std::max(time, 1 - m_frames[m_mark_in].ms + m_frames[m_mark_in - 1].ms);
+    }
+    else
+    {
+        time = std::clamp(time, 
+            1 - m_frames[m_mark_in].ms + m_frames[m_mark_in - 1].ms,
+            -1 - m_frames[m_mark_in].ms + m_frames[m_mark_in + 1].ms);
+        
+    }
+
+    m_frames[m_mark_in].ms += time;
+
+    return time != 0;
+}
+
+bool Replay::move_frames_range(SongTime_t time)
+{
+    if (m_mark_in == m_mark_out) {
+        return move_frame(time);
+    } 
+
+    if (m_mark_in == 0) {
+        time = std::clamp(time, 0, -1 - m_frames[m_mark_out].ms + m_frames[m_mark_out + 1].ms);
+    } 
+    else if (m_mark_out == m_frames.size() - 1) {
+        time = std::max(time, 1 - m_frames[m_mark_in].ms + m_frames[m_mark_in - 1].ms);
+    } 
+    else {
+        time = std::clamp(time, 
+            1 - m_frames[m_mark_in].ms + m_frames[m_mark_in - 1].ms, 
+            -1 - m_frames[m_mark_out].ms + m_frames[m_mark_out + 1].ms);
+    }
+
+    for (auto frame = m_frames.begin() + m_mark_in; frame != m_frames.begin() + m_mark_out + 1; ++frame) {
+        frame->ms += time;
+    }
+
+    return time != 0;
+}
+
+bool Replay::center_frame()
+{
+    if (m_mark_in >= m_frames.size() - 1 or m_mark_in < 0) {
+        return false;
+    }
+
+
+    if (m_mark_in == 0) {
+        SongTime_t new_time = m_frames[m_mark_in + 1].ms / 2;
+        new_time = std::clamp(new_time, 1, m_frames[m_mark_in + 1].ms - 1);
+        m_frames[m_mark_in].ms = new_time;
+    } else {
+        SongTime_t new_time = (m_frames[m_mark_in - 1].ms + m_frames[m_mark_in + 1].ms) / 2;
+        new_time = std::clamp(new_time, m_frames[m_mark_in - 1].ms + 1, m_frames[m_mark_in + 1].ms - 1);
+        m_frames[m_mark_in].ms = new_time;
+    }
+    return true;
+}
+
+bool Replay::center_frames_range()
+{
+    if (m_mark_in == m_mark_out) {
+        return center_frame();
+    } 
+    I64 size = m_mark_out - m_mark_in;
+
+    if (m_mark_in == 0) {
+        SongTime_t difference = m_frames[m_mark_out + 1].ms - m_frames[m_mark_in].ms;
+        if (difference < size) return 0;
+
+        int i = 1;
+        double base = (double)difference / (size + 1);
+
+        for (auto frame = m_frames.begin() + m_mark_in + 1; frame != m_frames.begin() + m_mark_out + 1; ++frame) {
+            frame->ms = m_frames[m_mark_in].ms + base * i;
+            ++i;
+        }
+    } 
+    else if (m_mark_out == m_frames.size() - 1) {
+        SongTime_t difference = m_frames[m_mark_out].ms - m_frames[m_mark_in - 1].ms;
+        if (difference < size) return 0;
+
+        int i = 1;
+        double base = (double)difference / (size + 1);
```

Comment:
```suggestion
        double base = static_cast<double>(difference) / (size + 1);
```

Reply:
honestly I don't understand the meaning of changing cast to c++ style
it looks ugly IMO
i'm using c++ style casting only in dynamic or interpret casts

Reply:
fixed